File: src/main/java/magma/option/OptionIter.java
package magma.option;

import magma.list.Iter;

public class OptionIter<T> implements Iter<T> {
    private final Option<T> self;
    private boolean done;

    public OptionIter(Option<T> self) {
        this.self = self;
        done = !self.isSome();
    }

    @Override
    public boolean hasNext() {
        return !done;
    }

    @Override
    public T next() {
        done = true;
        return self.get();
    }
}

File: src/main/node/magma/option/OptionIter.ts
import Iter from "../list/Iter";
export default class OptionIter<T> implements Iter<T> {
    private readonly self: Option<T>;
    private done: boolean;

    OptionIter(self: Option<T>): public {
        // TODO
        // TODO
    }

    @Override
    hasNext(): boolean {
        return !done;
    }

    @Override
    next(): T {
        // TODO
        return self.get();
    }
}
